File: app/src/main/java/com/storehop/app/data/csv/CsvImportExport.kt
package com.storehop.app.data.csv

/**
 * CSV import / export — pure functions, no DB or DI dependency.
 *
 * Supports the documented Storehop format only (header row required).
 * Items header columns: name, category, stores, brand, notes, quantity,
 *   isStaple, isPriority. Only `name` is required; unknown columns are
 *   ignored for forward compat. `stores` is a comma-separated list,
 *   typically quoted: "Aldi,Lidl,Pingo Doce".
 * Categories header columns: name, icon. Only `name` is required.
 *
 * Repository layer ([com.storehop.app.data.repository.ImportExportRepository])
 * is responsible for turning `*CsvRow` payloads into actual entities and
 * resolving category / store names to ids. This file is intentionally
 * decoupled from Room.
 */

data class ItemCsvRow(
    val name: String,
    val category: String? = null,
    val storeNames: List<String> = emptyList(),
    val brand: String? = null,
    val notes: String? = null,
    val quantity: String? = null,
    val isStaple: Boolean = false,
    val isPriority: Boolean = false,
)

data class CategoryCsvRow(
    val name: String,
    val icon: String? = null,
)

data class ParsedItemCsv(
    val rows: List<ItemCsvRow>,
    /** Per-row error messages prefixed with the source line number, 1-based. */
    val errors: List<String>,
)

data class ParsedCategoryCsv(
    val rows: List<CategoryCsvRow>,
    val errors: List<String>,
)

private val ITEM_HEADERS = listOf(
    "name", "category", "stores", "brand", "notes", "quantity", "isStaple", "isPriority",
)
private val CATEGORY_HEADERS = listOf("name", "icon")

// ---- Parse ----------------------------------------------------------------

fun parseItemCsv(content: String): ParsedItemCsv {
    val rawRows = parseCsvLines(content)
    if (rawRows.isEmpty()) return ParsedItemCsv(emptyList(), emptyList())

    val header = rawRows.first().map { it.trim().lowercase() }
    val nameIdx = header.indexOf("name")
    if (nameIdx < 0) {
        return ParsedItemCsv(
            rows = emptyList(),
            errors = listOf("Header row must include a 'name' column."),
        )
    }
    fun idx(col: String) = header.indexOf(col).takeIf { it >= 0 }

    val catIdx = idx("category")
    val storesIdx = idx("stores")
    val brandIdx = idx("brand")
    val notesIdx = idx("notes")
    val qtyIdx = idx("quantity")
    val stapleIdx = idx("isstaple")
    val priorityIdx = idx("ispriority")

    val rows = mutableListOf<ItemCsvRow>()
    val errors = mutableListOf<String>()
    rawRows.drop(1).forEachIndexed { i, fields ->
        val lineNumber = i + 2 // header is line 1; first data row is line 2
        if (fields.all { it.isBlank() }) return@forEachIndexed // skip blank rows silently
        val name = fields.getOrNull(nameIdx)?.trim().orEmpty()
        if (name.isEmpty()) {
            errors += "Line $lineNumber: 'name' is required."
            return@forEachIndexed
        }
        rows += ItemCsvRow(
            name = name,
            category = catIdx?.let { fields.getOrNull(it)?.trim().orEmptyOrNull() },
            storeNames = storesIdx?.let { fields.getOrNull(it).orEmpty() }
                ?.split(",")
                ?.map { it.trim() }
                ?.filter { it.isNotEmpty() }
                ?: emptyList(),
            brand = brandIdx?.let { fields.getOrNull(it)?.trim().orEmptyOrNull() },
            notes = notesIdx?.let { fields.getOrNull(it)?.trim().orEmptyOrNull() },
            quantity = qtyIdx?.let { fields.getOrNull(it)?.trim().orEmptyOrNull() },
            isStaple = stapleIdx?.let { fields.getOrNull(it)?.parseBool() } ?: false,
            isPriority = priorityIdx?.let { fields.getOrNull(it)?.parseBool() } ?: false,
        )
    }
    return ParsedItemCsv(rows, errors)
}

fun parseCategoryCsv(content: String): ParsedCategoryCsv {
    val rawRows = parseCsvLines(content)
    if (rawRows.isEmpty()) return ParsedCategoryCsv(emptyList(), emptyList())

    val header = rawRows.first().map { it.trim().lowercase() }
    val nameIdx = header.indexOf("name")
    if (nameIdx < 0) {
        return ParsedCategoryCsv(
            rows = emptyList(),
            errors = listOf("Header row must include a 'name' column."),
        )
    }
    val iconIdx = header.indexOf("icon").takeIf { it >= 0 }

    val rows = mutableListOf<CategoryCsvRow>()
    val errors = mutableListOf<String>()
    rawRows.drop(1).forEachIndexed { i, fields ->
        val lineNumber = i + 2
        if (fields.all { it.isBlank() }) return@forEachIndexed
        val name = fields.getOrNull(nameIdx)?.trim().orEmpty()
        if (name.isEmpty()) {
            errors += "Line $lineNumber: 'name' is required."
            return@forEachIndexed
        }
        rows += CategoryCsvRow(
            name = name,
            icon = iconIdx?.let { fields.getOrNull(it)?.trim().orEmptyOrNull() },
        )
    }
    return ParsedCategoryCsv(rows, errors)
}

// ---- Serialize ------------------------------------------------------------

/** Emit one CSV string with header row + one row per [rows] entry. */
fun List<ItemCsvRow>.toItemsCsv(): String = buildString {
    append(ITEM_HEADERS.joinToString(","))
    append('\n')
    this@toItemsCsv.forEach { r ->
        append(
            listOf(
                r.name,
                r.category.orEmpty(),
                r.storeNames.joinToString(","),
                r.brand.orEmpty(),
                r.notes.orEmpty(),
                r.quantity.orEmpty(),
                r.isStaple.toString(),
                r.isPriority.toString(),
            ).joinToString(",") { escapeCsvField(it) },
        )
        append('\n')
    }
}

fun List<CategoryCsvRow>.toCategoriesCsv(): String = buildString {
    append(CATEGORY_HEADERS.joinToString(","))
    append('\n')
    this@toCategoriesCsv.forEach { r ->
        append(escapeCsvField(r.name))
        append(',')
        append(escapeCsvField(r.icon.orEmpty()))
        append('\n')
    }
}

// ---- Internal RFC-4180 parser ---------------------------------------------

/**
 * Split a CSV blob into a list of records, each a list of unescaped fields.
 * Handles \r\n and \n line endings, quoted fields with embedded commas /
 * newlines, and "" as a literal quote inside a quoted field.
 */
internal fun parseCsvLines(content: String): List<List<String>> {
    if (content.isEmpty()) return emptyList()
    val rows = mutableListOf<MutableList<String>>()
    var current = mutableListOf<String>()
    val field = StringBuilder()
    var inQuotes = false
    var i = 0
    val n = content.length

    fun pushField() { current.add(field.toString()); field.clear() }
    fun pushRow() {
        pushField()
        rows.add(current)
        current = mutableListOf()
    }

    while (i < n) {
        val c = content[i]
        when {
            inQuotes -> when {
                c == '"' && i + 1 < n && content[i + 1] == '"' -> { // escaped quote
                    field.append('"')
                    i++
                }
                c == '"' -> inQuotes = false
                else -> field.append(c)
            }
            c == '"' && field.isEmpty() -> inQuotes = true
            c == ',' -> pushField()
            c == '\r' -> {
                if (i + 1 < n && content[i + 1] == '\n') i++
                pushRow()
            }
            c == '\n' -> pushRow()
            else -> field.append(c)
        }
        i++
    }
    // Flush trailing field/row (no terminating newline)
    if (field.isNotEmpty() || current.isNotEmpty()) pushRow()
    // Drop entirely-empty trailing rows produced by a final newline
    return rows.filterNot { it.size == 1 && it[0].isEmpty() }
}

/** Quote a field if it contains a comma, double-quote, or any line break;
 *  double up embedded quotes per RFC 4180. */
internal fun escapeCsvField(s: String): String {
    val needsQuoting = s.any { it == ',' || it == '"' || it == '\n' || it == '\r' }
    return if (!needsQuoting) s
    else "\"" + s.replace("\"", "\"\"") + "\""
}

private fun String?.orEmptyOrNull(): String? = this?.takeIf { it.isNotEmpty() }
private fun String.parseBool(): Boolean = trim().equals("true", ignoreCase = true)

File: ios/Storehop/Data/CSV/CsvImportExport.swift
import Foundation

/// CSV import / export — pure functions, no DB or DI dependency. Mirrors
/// Android's `data/csv/CsvImportExport.kt`.
///
/// Items header columns: name, category, stores, brand, notes, quantity,
///   isStaple, isPriority. Only `name` is required; unknown columns are
///   ignored for forward compat. `stores` is a comma-separated list,
///   typically quoted: "Aldi,Lidl,Pingo Doce".
/// Categories header columns: name, icon. Only `name` is required.

struct ItemCsvRow: Equatable {
    let name: String
    let category: String?
    let storeNames: [String]
    let brand: String?
    let notes: String?
    let quantity: String?
    let isStaple: Bool
    let isPriority: Bool

    init(
        name: String,
        category: String? = nil,
        storeNames: [String] = [],
        brand: String? = nil,
        notes: String? = nil,
        quantity: String? = nil,
        isStaple: Bool = false,
        isPriority: Bool = false
    ) {
        self.name = name
        self.category = category
        self.storeNames = storeNames
        self.brand = brand
        self.notes = notes
        self.quantity = quantity
        self.isStaple = isStaple
        self.isPriority = isPriority
    }
}

struct CategoryCsvRow: Equatable {
    let name: String
    let icon: String?
}

struct ParsedItemCsv {
    let rows: [ItemCsvRow]
    /// Per-row error messages prefixed with the source line number, 1-based.
    let errors: [String]
}

struct ParsedCategoryCsv {
    let rows: [CategoryCsvRow]
    let errors: [String]
}

private let ITEM_HEADERS = [
    "name", "category", "stores", "brand", "notes", "quantity", "isStaple", "isPriority",
]
private let CATEGORY_HEADERS = ["name", "icon"]

// MARK: - Parse

func parseItemCsv(_ content: String) -> ParsedItemCsv {
    let rawRows = parseCsvLines(content)
    if rawRows.isEmpty { return ParsedItemCsv(rows: [], errors: []) }

    let header = rawRows[0].map { $0.trimmed.lowercased() }
    guard let nameIdx = header.firstIndex(of: "name") else {
        return ParsedItemCsv(rows: [], errors: ["Header row must include a 'name' column."])
    }
    let catIdx = header.firstIndex(of: "category")
    let storesIdx = header.firstIndex(of: "stores")
    let brandIdx = header.firstIndex(of: "brand")
    let notesIdx = header.firstIndex(of: "notes")
    let qtyIdx = header.firstIndex(of: "quantity")
    let stapleIdx = header.firstIndex(of: "isstaple")
    let priorityIdx = header.firstIndex(of: "ispriority")

    var rows: [ItemCsvRow] = []
    var errors: [String] = []
    for (i, fields) in rawRows.dropFirst().enumerated() {
        let lineNumber = i + 2  // header = line 1; first data row = line 2
        if fields.allSatisfy({ $0.isBlank }) { continue }
        let name = fields.indexed(nameIdx)?.trimmed ?? ""
        if name.isEmpty {
            errors.append("Line \(lineNumber): 'name' is required.")
            continue
        }
        rows.append(ItemCsvRow(
            name: name,
            category: catIdx.flatMap { fields.indexed($0)?.trimmed.nilIfEmpty },
            storeNames: storesIdx.flatMap { fields.indexed($0) }
                .map { $0.split(separator: ",").map { String($0).trimmed }.filter { !$0.isEmpty } }
                ?? [],
            brand: brandIdx.flatMap { fields.indexed($0)?.trimmed.nilIfEmpty },
            notes: notesIdx.flatMap { fields.indexed($0)?.trimmed.nilIfEmpty },
            quantity: qtyIdx.flatMap { fields.indexed($0)?.trimmed.nilIfEmpty },
            isStaple: stapleIdx.flatMap { fields.indexed($0)?.parsedBool } ?? false,
            isPriority: priorityIdx.flatMap { fields.indexed($0)?.parsedBool } ?? false
        ))
    }
    return ParsedItemCsv(rows: rows, errors: errors)
}

func parseCategoryCsv(_ content: String) -> ParsedCategoryCsv {
    let rawRows = parseCsvLines(content)
    if rawRows.isEmpty { return ParsedCategoryCsv(rows: [], errors: []) }

    let header = rawRows[0].map { $0.trimmed.lowercased() }
    guard let nameIdx = header.firstIndex(of: "name") else {
        return ParsedCategoryCsv(rows: [], errors: ["Header row must include a 'name' column."])
    }
    let iconIdx = header.firstIndex(of: "icon")

    var rows: [CategoryCsvRow] = []
    var errors: [String] = []
    for (i, fields) in rawRows.dropFirst().enumerated() {
        let lineNumber = i + 2
        if fields.allSatisfy({ $0.isBlank }) { continue }
        let name = fields.indexed(nameIdx)?.trimmed ?? ""
        if name.isEmpty {
            errors.append("Line \(lineNumber): 'name' is required.")
            continue
        }
        rows.append(CategoryCsvRow(
            name: name,
            icon: iconIdx.flatMap { fields.indexed($0)?.trimmed.nilIfEmpty }
        ))
    }
    return ParsedCategoryCsv(rows: rows, errors: errors)
}

// MARK: - Serialize

extension Array where Element == ItemCsvRow {
    func toItemsCsv() -> String {
        var out = ITEM_HEADERS.joined(separator: ",") + "\n"
        for r in self {
            let cols = [
                r.name,
                r.category ?? "",
                r.storeNames.joined(separator: ","),
                r.brand ?? "",
                r.notes ?? "",
                r.quantity ?? "",
                r.isStaple ? "true" : "false",
                r.isPriority ? "true" : "false",
            ]
            out += cols.map(escapeCsvField).joined(separator: ",") + "\n"
        }
        return out
    }
}

extension Array where Element == CategoryCsvRow {
    func toCategoriesCsv() -> String {
        var out = CATEGORY_HEADERS.joined(separator: ",") + "\n"
        for r in self {
            out += escapeCsvField(r.name) + "," + escapeCsvField(r.icon ?? "") + "\n"
        }
        return out
    }
}

// MARK: - Internal RFC-4180 parser

/// Split a CSV blob into a list of records, each a list of unescaped fields.
/// Handles \r\n and \n line endings, quoted fields with embedded commas /
/// newlines, and "" as a literal quote inside a quoted field.
func parseCsvLines(_ content: String) -> [[String]] {
    if content.isEmpty { return [] }
    var rows: [[String]] = []
    var current: [String] = []
    var field = ""
    var inQuotes = false
    let chars = Array(content)
    var i = 0
    let n = chars.count

    func pushField() {
        current.append(field)
        field = ""
    }
    func pushRow() {
        pushField()
        rows.append(current)
        current = []
    }

    while i < n {
        let c = chars[i]
        if inQuotes {
            if c == "\"" && i + 1 < n && chars[i + 1] == "\"" {
                field.append("\"")
                i += 1
            } else if c == "\"" {
                inQuotes = false
            } else {
                field.append(c)
            }
        } else {
            if c == "\"" && field.isEmpty {
                inQuotes = true
            } else if c == "," {
                pushField()
            } else if c == "\r" {
                if i + 1 < n && chars[i + 1] == "\n" { i += 1 }
                pushRow()
            } else if c == "\n" {
                pushRow()
            } else {
                field.append(c)
            }
        }
        i += 1
    }
    if !field.isEmpty || !current.isEmpty { pushRow() }
    // Drop entirely-empty trailing rows produced by a final newline.
    return rows.filter { !($0.count == 1 && $0[0].isEmpty) }
}

/// Quote a field if it contains a comma, double-quote, or any line break;
/// double up embedded quotes per RFC 4180.
func escapeCsvField(_ s: String) -> String {
    let needsQuoting = s.contains(",") || s.contains("\"") || s.contains("\n") || s.contains("\r")
    if !needsQuoting { return s }
    return "\"" + s.replacingOccurrences(of: "\"", with: "\"\"") + "\""
}

// MARK: - Helpers

private extension String {
    var trimmed: String { trimmingCharacters(in: .whitespacesAndNewlines) }
    var isBlank: Bool { trimmed.isEmpty }
    var nilIfEmpty: String? { isEmpty ? nil : self }
    var parsedBool: Bool { trimmed.lowercased() == "true" }
}

private extension Array where Element == String {
    /// Bounds-safe index access mirroring Kotlin's `getOrNull(index)`.
    func indexed(_ index: Int) -> String? {
        guard index >= 0 && index < count else { return nil }
        return self[index]
    }
}
